crates/net/eth-wire/src/types/message.rs

```diff
@@ -237,6 +237,99 @@ impl EthMessage {
+    ///
+    /// # Arguments
+    ///
+    /// * `approx` - The approximate size to allocate in memory
+    /// * `size` - The maximum allowed size of the data.
+    ///
+    /// # Returns
+    ///
+    /// * `Ok(Vec<u8>)` - A vector containing the encoded data if it is within the size limit.
+    /// * `Err(alloy_rlp::Error)` - An error if the encoded data exceeds the size limit.
+    fn encode_max(
+        &self,
+        approx: usize,
+        size: usize,
+    ) -> alloy_rlp::Result<Vec<u8>, alloy_rlp::Error>;
+
+    /// Encodes the data, truncating the encoded message if it exceeds the specified size limit.
+    /// The resulting data will fit within the limit but may be incomplete if truncation occurs.
+    ///
+    /// When using this method with collections or "vec-like" structures, implementors should take
+    /// care to ensure the truncated encoding remains valid and interpretable, albeit potentially
+    /// incomplete. For example, truncating a list of items might result in the list containing
+    /// fewer items than intended, but each item in the list should remain individually intact
+    /// and decodable.
+    ///
+    /// # Arguments
+    ///
+    /// * `approx` - The approximate size to allocate in memory
+    /// * `limit` - The maximum allowed size of the encoded data.
+    ///
+    /// # Returns
+    ///
+    /// A vector containing the encoded data, truncated to fit within the size limit if necessary.
+    /// Users of this method should be aware that the truncated data may not represent the complete
+    /// original dataset.
+    fn encode_truncate(&self, approx: usize, limit: usize) -> Vec<u8>;
+}
+
+impl<T: Encodable> EncodableExt for Vec<T> {
+    fn encode_max(
+        &self,
+        approx: usize,
+        limit: usize,
+    ) -> alloy_rlp::Result<Vec<u8>, alloy_rlp::Error> {
+        let mut buffer = Vec::with_capacity(approx);
+        let mut header = Header { list: true, payload_length: 0 };
+        for item in self {
+            header.payload_length += item.length();
+        }
+        header.encode(&mut buffer);
+
+        for item in self {
+            item.encode(&mut buffer);
+            if buffer.len() > limit {
+                return Err(alloy_rlp::Error::Custom("Size limit exceeded"));
+            }
+        }
+        Ok(buffer)
+    }
+
+    fn encode_truncate(&self, approx: usize, limit: usize) -> Vec<u8> {
+        let mut buffer = Vec::with_capacity(approx);
+        let mut header = Header { list: true, payload_length: 0 };
+        for item in self {
+            header.payload_length += item.length();
+        }
+        header.encode(&mut buffer);
+        for item in self {
+            let current_len = buffer.len();
+            item.encode(&mut buffer);
+            if buffer.len() > limit {
+                buffer.truncate(current_len);
+                break;
+            }
+        }
+        buffer
```

Comment:
please `shrink_to_fit` the buffer here. otherwise the allocated memory for the buffer will stay the approx size, even if it's filled with viewer elements! that would result in a logic bug if you are packing a response based on the length (number of elements) in the vector.